### PR TITLE
database: fix reindex with uninstalled deps

### PR DESCRIPTION
**Pull request by haampie**
When a prefix of a dep is removed, and the db is reindexed, it is added
through the dependent, but until now it incorrectly listed the spec as
'installed'.

It also fixes an issue when the same spec was added multiple times, it would
always be marked installed, even though it was not.

Lastly it fixes some edge cases where removing an uninstalled, non-external
spec errored if it had never been installed in the database in the first place
(e.g. prefix of some root spec is gone, it's added back on spack reindex
through the old index.json -- but due to a bug you would not be able to
remove this missing entry from the database.)

### CODE REVIEW

File: lib/spack/spack/database.py [hunk 1]
```diff
@@ -940,22 +940,15 @@ def _construct_from_directory_layout(self, directory_layout, old_data):
...
-                    if install_check:
-                        kwargs = {
-                            'spec': entry.spec,
-                            'directory_layout': layout,
-                            'explicit': entry.explicit,
-                            'installation_time': entry.installation_time  # noqa: E501
-                        }
-                        self._add(**kwargs)
-                        processed_specs.add(entry.spec)
+                    layout = None if entry.spec.external else spack.store.layout
```
Comment (haampie) [inline lib/spack/spack/database.py:943]: this is somewhat duplicate logic, _add takes care of it anways (and probs better), hence refactor.

File: lib/spack/spack/database.py [hunk 2]
```diff
@@ -1099,24 +1092,28 @@ def _add(
...
-            if not spec.external and directory_layout:
-                path = directory_layout.path_for_spec(spec)
-                if path in self._installed_prefixes:
-                    raise Exception("Install prefix collision.")
-                try:
-                    directory_layout.check_installed(spec)
-                    installed = True
-                except DirectoryLayoutError as e:
+        # Make sure the directory layout agrees whether the spec is installed
+        installed = bool(spec.external)
```
Comment (haampie) [inline lib/spack/spack/database.py]: so, here we retrieve installed-ness from the directory layout, also when the spec had been in the db before.

File: lib/spack/spack/database.py [hunk 3]
```diff
@@ -1146,8 +1143,8 @@ def _add(
         else:
             # If it is already there, mark it as installed and update
             # installation time
-            self._data[key].installed = True
-            self._data[key].installation_time = _now()
+            self._data[key].installed = installed
```
Comment (haampie) [inline lib/spack/spack/database.py:1145]: This was a bug: adding the same (uninstalled) record twice doesn't imply it's magically on the filesystem.

File: lib/spack/spack/database.py [hunk 4]
```diff
@@ -1099,24 +1092,28 @@ def _add(
...
-                        'Dependency missing: may be deprecated or corrupted:',
-                        path, str(e))
+        # Make sure the directory layout agrees whether the spec is installed
+        if not spec.external and directory_layout:
+            path = directory_layout.path_for_spec(spec)
+            installed = False
+            try:
+                directory_layout.ensure_installed(spec)
+                installed = True
                 self._installed_prefixes.add(path)
```
Comment (haampie) [inline lib/spack/spack/database.py:1102]: For those reviewing and out of the loop: the invariant of `_installed_prefixes` is: for every path in the set, there exists 1 install record in the database with `rec.installed = True` and `rec.spec.external = False`.

File: lib/spack/spack/database.py [hunk 5]
```diff
@@ -1213,7 +1209,7 @@ def _remove(self, spec):
 
         # This install prefix is now free for other specs to use, even if the
         # spec is only marked uninstalled.
-        if not rec.spec.external:
+        if not rec.spec.external and rec.installed:
```
Comment (haampie) [inline lib/spack/spack/database.py:1212]: This was also a bug: only part of the invariant was checked.

It was only hit thanks to the faulty test described in https://github.com/spack/spack/issues/28779: this test added an entry to the database, but it was not considered installed on the filesystem (spec passed to database did not correspond to spec on filesystem). So it ends up as registering an entry in the database with `rec.installed==False`, `rec.spec.external==False` and `rec.ref_count==0`...

An actual test could be: a package that somehow destructs its own install prefix after installation and before being registered in the database. Although it's quite edge-case'y and unclear what expected behavior should be.

Comment (haampie) [reply lib/spack/spack/database.py:1212]: This is now covered by a test.

File: lib/spack/spack/database.py [hunk 6]
```diff
@@ -1144,10 +1141,9 @@ def _add(
             new_spec._full_hash = spec._full_hash
 
         else:
-            # If it is already there, mark it as installed and update
-            # installation time
-            self._data[key].installed = True
-            self._data[key].installation_time = _now()
+            # It is already in the database
+            self._data[key].installed = installed
+            self._data[key].installation_time = installation_time
```
Comment (becker33) [inline lib/spack/spack/database.py]: This needs to only update the installation time if the `installed` value changed.

```suggestion
            new_install = self._data[key].installed != installed
            self._data[key].installed = installed
            if new_install:
                self._data[key].installation_time = installation_time
```

File: lib/spack/spack/database.py [hunk 7]
```diff
@@ -1099,24 +1092,28 @@ def _add(
...
+                msg = ("{0} is being {1} in the database with prefix {2}, "
+                       "but this directory does not contain an installation of "
+                       "the spec, due to: {3}")
+                action = "updated" if key in self._data else "registered"
+                tty.warn(msg.format(spec.short_spec, action, path, str(e)))
+        elif spec.external_path:
+            path = spec.external_path
+            installed = True
+        else:
+            path = None
```
Comment (becker33) [inline lib/spack/spack/database.py:1113]: I think this is just for compatibility with spack v0.8 at this point, I wonder if we can remove it soon.

**Review by becker33 — CHANGES_REQUESTED** [review]
Mostly LGTM, one bug that needs fixing with the installation time.

File: lib/spack/spack/database.py (hunk 6 above)
Comment (haampie) [reply lib/spack/spack/database.py]: I think this codepath is hit for overwrite installs (like develop specs), in which case you'd want to "touch" the record to reflect the overwrite time as installation_time, right?

If it goes from installed -> not installed, then probably you don't want the installation_time to be modified.

Comment (becker33) [reply lib/spack/spack/database.py]: I'm pretty sure this code path is also hit for

```
spack install bar
spack install foo^bar  # hits this code path for bar
```

But you're right, we need something cleverer than my suggestion for the overwrite case.

Comment (haampie) [reply lib/spack/spack/database.py]: Hm okay, right now the behavior is effectively unchanged if the install time is not passed as an argument. Maybe getting install times right could be done in a follow up PR? Seems like a few code paths have to be considered... (install-as-dependency, overwrite, re-registering through reindex...)

Comment (becker33) [reply lib/spack/spack/database.py]: We could just stat the `spec.prefix` directory for its creation time, and only update the installation time for previously installed packages if that time is greater than the existing installation time.

```suggestion
            set_installation_time = installed and not self._data[key].installed
            if self._data[key].installed and installed:
                # On UNIX, creation times are really metadata modification times.
                # shouldn't make a difference in this case
                prefix_creation_time = os.stat(spec.prefix).st_ctime
                if prefix_creation_time > self._data[key].installation_time:
                    set_installation_time = True
                   
            self._data[key].installed = installed
            if set_installation_time:
                self._data[key].installation_time = installation_time
```

Comment (becker33) [reply lib/spack/spack/database.py]: But I agree that this PR doesn't make this particular issue worse, so if you decide it needs to be pushed to a next PR I'm fine with that.

Comment (haampie) [reply lib/spack/spack/database.py]: I'm a bit reluctant to add this, considering `installation_time` is relied on in `installer.py`:

```
            if installed_in_db and (
                    dep.dag_hash() not in request.overwrite or
                    rec.installation_time > request.overwrite_time):
```

and I can already foresee hard to debug issues with distributed builds etc. What I'd like to have in this pr is clear bug fixes and backport them. Maybe it's best to stick to `_now()` like it was and fix install times separately but only for 0.18.

**Review by becker33 — APPROVED** [review]

**Comment by becker33** [comment]
> and ensure ensure only ensures

:)
